02_bubble/marchitect/main.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+	"math/big"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	// n is the number of items in the int array
+	n := len(s)
+	// make a copy of the original int array s
+	result := make([]int, n)
+	copy(result, s)
+	// maybe necessary
+	if n < 2 {
+		return result
+	}
+	// set swapped to true
+	swapped := true
+	// in Go, use keyword for to act as a while loop
+	for swapped {
+		// set swapped to false
+		swapped = false
+		// iterate through all of the elements in our list
+		for i := 1; i < n; i++ {
```

Comment:
The inner loop usually goes from `n` down. Reason: after the first outer iteration the largest element has "bubbled up" to the slice position with the higher index.
Our reference solution looks like this - feel free to use, adjust or ignore:
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			// If no elements were swapped, the slice is sorted so we can stop here
			return
		}
	}
```